File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/reader/ReaderFieldType.java
package ru.tinkoff.kora.json.annotation.processor.reader;

import jakarta.annotation.Nullable;
import ru.tinkoff.kora.json.annotation.processor.KnownType;

import javax.lang.model.type.TypeMirror;

public interface ReaderFieldType {

    enum JsonValueType {
        VALUE,
        NULLABLE,
        UNDEFINED
    }

    @Nullable
    JsonValueType jsonValueType();

    default boolean isJsonNullable() {
        return jsonValueType() != null
               && (jsonValueType() == ReaderFieldType.JsonValueType.VALUE || jsonValueType() == ReaderFieldType.JsonValueType.NULLABLE);
    }

    default boolean isJsonUndefined() {
        return jsonValueType() != null
               && (jsonValueType() == ReaderFieldType.JsonValueType.VALUE || jsonValueType() == JsonValueType.UNDEFINED);
    }

    TypeMirror typeMirror();

    record KnownTypeReaderMeta(KnownType.KnownTypesEnum knownType, TypeMirror typeMirror, JsonValueType jsonValueType) implements ReaderFieldType {}

    record UnknownTypeReaderMeta(TypeMirror typeMirror, JsonValueType jsonValueType) implements ReaderFieldType {}
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/reader/ReaderFieldType.kt
package ru.tinkoff.kora.json.ksp.reader

import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.TypeName
import ru.tinkoff.kora.json.ksp.KnownType

sealed interface ReaderFieldType {

    val typeName: TypeName

    val type: KSType

    val isJsonNullable: Boolean

    data class KnownTypeReaderMeta(
        override val type: KSType,
        override val typeName: TypeName,
        override val isJsonNullable: Boolean,
        val knownType: KnownType.KnownTypesEnum
    ) : ReaderFieldType

    data class UnknownTypeReaderMeta(
        override val type: KSType,
        override val typeName: TypeName,
        override val isJsonNullable: Boolean
    ) : ReaderFieldType
}
